03 Writing Algorithms/20 Historical Data/04 Asset Classes/09 Index Options/99 Examples.html

```diff
@@ -1 +1,200 @@
+            var residual = history.Select(x =&gt; CalculateSpread(x[_spxwContract].Close, x[_spxContract].Close, _cointegrationVector)).ToList();
+            _spreadMean = residual.Average();
+            _spreadSd = Convert.ToDecimal(Math.Sqrt(residual.Sum(x =&gt; Math.Pow((double)(x - _spreadMean), 2) / residual.Count)));
+        }
+    }
+    
+    private decimal[] CalculateCointegrationVector(IEnumerable&lt;DataDictionary&lt;QuoteBar&gt;&gt; history)
+    {
+        // Use log price to eliminate the compounding effect.
+        var y = history.Select(x =&gt; Math.Log((double)x[_spxwContract].Close)).ToArray();
+        var x1 = history.Select(x =&gt; Math.Log((double)x[_spxContract].Close)).ToArray();
+
+        // Create a matrix for the regression analysis.
+        var n = new[] { y.Length, x1.Length }.Min();
+        var designMatrix = new double[n][];
+        for (int i = 0; i &lt; n; i++)
+        {
+            designMatrix[i] = new[] { x1[i] };
+        }
+
+        try
+        {
+            // Perform regression using MathNet.Numerics
+            var coefficients = MultipleRegression.QR(designMatrix, y.TakeLast(n).ToArray(), intercept: true);
+
+            // The coefficients array will contain the intercept and the coefficients for each independent variable.
+            // Format: [intercept, x1]
+            return new[] { (decimal)coefficients[0], (decimal)coefficients[1] };
+        }
+        catch
+        {
+            return _cointegrationVector;
+        }
+    }
+
+    private decimal CalculateSpread(decimal y, decimal x1, decimal[] cointegrationVector)
+    {
+        // Using the cointegration vector to calculate the spread.
+        return LogPrice(y) - (cointegrationVector[0] + LogPrice(x1) * cointegrationVector[1]);
+    }
+
+    private decimal LogPrice(decimal price)
+    {
+        return Convert.ToDecimal(Math.Log((double)price));
+    }
+}</pre>
+    <pre class="python">from sklearn.linear_model import LinearRegression
+
+class IndexOptionHistoricalDataAlgorithm(QCAlgorithm):
+    _threshold = 2
+    _coef = 0
+    _intercept = 0
+    _mean_spread = 0
+    _sd_spread = 1
+    _spx_contract = None
+    _spxw_contract = None
+
+    def initialize(self) -&gt; None:
+        self.set_start_date(2025, 1, 1)
+        self.set_end_date(2025, 2, 1)
+        self.set_cash(10000000)
+        
+        # Request SPX and SPXW options to analyze their relationship and trade.
+        self._spx = self.add_index_option("SPX").symbol
+        self._spxw = self.add_index_option("SPX", "SPXW").symbol
+
+        # Select the Index Options to analyze and trade by week.
+        self.schedule.on(self.date_rules.week_start(), self.time_rules.at(9, 15), self.select_contracts)
+
+    def select_contracts(self) -&gt; None:
+        # Obtain the SPX ATM call contract, since it is the most liquid to trade with.
+        spx_contracts = [x for x in self.option_chain(self._spx) if x.expiry &lt; self.time + timedelta(30)]
```

Comment:
We shouldn't use `self.option_chain` if we are not adding contracts manually because the contracts might not been selected by the default filter.